src/mlpack/methods/ann/loss_functions/cosine_embedding_loss.hpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * @file cosine_embedding_loss.hpp
+ * @author Kartik Dutt
+ *
+ * Definition of the Cosine Embedding loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_COSINE_EMBEDDING_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * Cosine Embeddings Loss function is used for measuring whether two inputs are
+ * similar or dissimilar, using using the cosine distance, and is typically used
+ * for learning nonlinear embeddings or semi-supervised learning.
+ *
+ * @f{eqnarray*}{
+ * f(x) = 1 - cos(x1, x2) , for y = 1
+ * f(x) = max(0, cos(x1, x2) - margin) , for y = -1
+ * @f}
+ * 
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ */
+template <
+    typename InputDataType = arma::mat,
+    typename OutputDataType = arma::mat
+>
+class CosineEmbeddingLoss
+{
+ public:
+  /**
+   * Create the CosineEmbeddingLoss object.
+   * 
+   * @param margin Increases cosine distance in case of dissimilarity.
+   *               Refer definition of cosine-embedding-loss above.
+   * @param takeMean Boolean variable to specify whether to take mean or not.
+   *                 Specifies reduction method i.e. sum or mean corresponding
+   *                 to 0 and 1 respectively. Default value = 0.
+   */
+  CosineEmbeddingLoss(const double margin = 0.0, const bool takeMean = false);
+
+  /**
+   * Ordinary feed forward pass of a neural network.
+   *
+   * @param x1 Input data used for evaluating the given function.
+   * @param x2 Input data used for evaluating the given function.
+   * @param y Input data used to determine whether to calculate
+   *          cosine similarity or dissimilarity. It should only
+   *          contain values equal to 1 or -1.
+   */
+  template<
+    typename FirstTensor,
```

Comment:
Naming nomenclature, I would have liked to use variable names that are more intuitive.

Reply:
I think you're right; Maybe something like FirstInputType SecondInputType TargetType that differentiates between x_n and y?

Reply:
Hi @birm, I will make the changes for the same.